### DISCUSSION TITLE
Using Typescript interfaces with Prisma

### DISCUSSION THREAD

**Discussion by troncoso** [prisma/prisma#14112] [Q&A]
I've searched around and this has been somewhat asked, but most answers are just to use schema features.

I have model interfaces such as:

```typescript
export interface User {
    id?: number;
    authId?: string;
    firstName?: string;
    lastName?: string;
    markets?: Market[];
}

export interface Markets {
    id?: number;
    name?: string;
}
```

These are in a shared package accessible by both my client and server code. This lets me synchronize the request/response data that is passed between the two. Is there a sensible way that Prisma will let my queries adhere to these interfaces? I would like `prisma.users.findFirst` to uphold my User interface contract.

EDIT

I did some more digging and found these related issues:

https://github.com/prisma/prisma/issues/5273
https://github.com/prisma/prisma/issues/3929

Not that I'm necessarily trying to build my own CRUD services on the backend, but the same core issue kind of lends to all of these problems.

It seems like its quite a disadvantage that you can't *easily* separate your entity format from your database format. So now to send data to the client, I would have to format my database to match how I'm sending it to them. This breaks down very quickly as new use cases arise.

Sorry, I'm new to Prisma. I'm trying to switch to it from Typeorm because I'm sick of the bugs and incomplete docs/api, but Prisma just doesn't seem that manageable/scaleable without being able to control the types of the crud operations.

Am I missing something here?

**Comment by austincrim** [comment]
Hey @troncoso 👋,

The interfaces you use for requests/responses and your database schema are separate things and it is generally advisable to type them separately as well.

You can build your own request/response types (using TypeScript utility types) based on the generated Prisma types while still keeping them as separate entities.

If you have ideas on how we can improve this experience, leaving feedback in the issues you linked would be a great place to start!

**Reply by troncoso** [reply]
Thanks for the response @austincrim . Do you have an example of using the generated types like you mentioned? That would be useful to see.

**Reply by austincrim** [reply]
There are lots of ways to accomplish what I'm referencing, but here's one example:
```ts
import type { User } from '@prisma/client'

type CreateUser = Omit<User, 'id' | 'createdAt' | 'anythingElseYouWantToExclude'>
```

You can accomplish a lot with the [built-in utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
